File: tests/enums/arithmetic.swift
public enum ArithmeticExpression {
    case number(Double)
    indirect case addition(ArithmeticExpression, ArithmeticExpression)
    indirect case multiplication(ArithmeticExpression, ArithmeticExpression)
}

public func literal(_ value: Double) -> ArithmeticExpression {
	return .number(value)
}

public func add(_ left: ArithmeticExpression, _ right: ArithmeticExpression) -> ArithmeticExpression {
	return .addition(left, right)
}

public func multiply(_ left: ArithmeticExpression, _ right: ArithmeticExpression) -> ArithmeticExpression {
	return .multiplication(left, right)
}

public func eval(_ expression: ArithmeticExpression) -> Double {
	switch expression {
		case .number(let value):
			return value
		case .addition(let l, let r):
			return eval(l) + eval(r)
		case .multiplication(let l, let r):
			return eval(l) * eval(r)
	}
}

public func silly(_ expression: ArithmeticExpression) -> Double {
	switch expression {
		case .number(var value):
			value += 10
			return value
		case .addition(var l, _):
			l = literal(10)
			return eval(l)
		default:
			return 0
	}
}

File: tests/enums/arithmetic.js
export function literal(value) {
  return [0, value];
}

function ArithmeticExpression$copy(source) {
  return source[0] === 2 ? [2, ArithmeticExpression$copy(source[1]), ArithmeticExpression$copy(source[2])] : source[0] === 1 ? [1, ArithmeticExpression$copy(source[1]), ArithmeticExpression$copy(source[2])] : source.slice();
}

export function add(left, right) {
  return [1, ArithmeticExpression$copy(left), ArithmeticExpression$copy(right)];
}
export function multiply(left, right) {
  return [2, ArithmeticExpression$copy(left), ArithmeticExpression$copy(right)];
}
export function eval(expression) {
  var $match = expression;

  if ($match[0] === 0) {
    const value = $match[1];
    return value;
  } else if ($match[0] === 1) {
    const l = ArithmeticExpression$copy($match[1]);
    const r = ArithmeticExpression$copy($match[2]);
    return eval(l) + eval(r);
  } else if ($match[0] === 2) {
    const l = ArithmeticExpression$copy($match[1]);
    const r = ArithmeticExpression$copy($match[2]);
    return eval(l) * eval(r);
  }
}
export function silly(expression) {
  var $match = expression;

  if ($match[0] === 0) {
    let value = $match[1];
    value += 10;
    return value;
  } else if ($match[0] === 1) {
    let l = ArithmeticExpression$copy($match[1]);
    l = literal(10);
    return eval(l);
  } else {
    return 0;
  }
}